Update NServiceBus overview page (#4385)
Particular/docs.particular.net: platform/nservicebus.include.md

NServiceBus is the heart of a distributed system and the Particular Service Platform. It helps create systems that are scalable, reliable, and flexible.

> At its core, NServiceBus works by routing _messages_ between _endpoints_. [Messages](/nservicebus/concept-overview.md#message) are plain C# classes that contain meaningful data for the business process that is being modeled.

Reviewer: Do we need to provide a link to explain what is endpoint?

Reviewer 2: There is one at the top of the next paragraph. The idea was to define a message, show one, then define an endpoint, and show one.

```csharp

> public class ProcessOrder : IMessage

Reviewer 3: I would consider showing unobtrusive here, which should probably be more strongly advertised as the default throughout our docs. I know for a fact that some developers find inheriting from a base type for a message highly off-putting, and it may not be obvious that it's optional.

```suggestion
public class ProcessOrder
```

{
    public int OrderId { get; set; }
}
```

[Endpoints](/nservicebus/concept-overview.md#endpoint) are logical entities that send and/or receive messages.

```csharp
// Sending endpoint
await endpoint.Send(new ProcessOrder { OrderId = 15 });

// Receiving endpoint
public class ProcessOrderHandler : IHandleMessages<ProcessOrder>
{
    public async Task Handle(ProcessOrder messsage, IMessageHandlerContext context)
    {
        // Do something with ProcessOrder message
    }
}
```

Endpoints can be running in different processes, on different machines, even at different times. NServiceBus makes sure that each message reaches its intended destination and is processed.

NServiceBus accomplishes this by providing an abstraction over [existing queuing technologies](/transports/). While it's possible to work directly with queuing systems, NServiceBus provides extra features to make applications more reliable and scalable.

## Reliable

> NServiceBus offers different ways of ensuring information is not lost due to failures in a system. First, NServiceBus provides native transaction support (e.g. DTC) for the underlying queuing technologies that support it, as well as its own transaction guarantees through the implementation of the [Outbox pattern](/nservicebus/outbox).

Reviewer 3: > e.g. DTC

This may not be the best thing to mention here. Ultimately, NServiceBus will use transactions appropriately, and in many cases they do not elevate to the DTC. E.g. when using a SQL DB for transport, persistence, and business data. Those that do elevate to the DTC would probably only involve MSMQ and SQLT, which is a scenario which is probably diminishing in importance.

In other cases, NServiceBus has [built-in recoverability](/nservicebus/recoverability) that can automatically adapt to common failures in a system. For intermittent failures, such as network outages, messages can be retried at regular intervals. For more serious errors, messages are set aside in a separate error queue so that they can be investigated at a later time without impacting the overall performance of the system.

## Scalable

NServiceBus is designed to handle a large number of messages. Endpoints are configured for high performance by default, handling multiple messages in parallel. Depending on the workload, the number of messages handled concurrently can be [increased to improve message throughput](/nservicebus/operations/tuning.md).

> In high volume scenarios, where there are more messages being produced than a single endpoint can handle, each endpoint can be [scaled out across multiple instances](/transports/scale-out.md) running on different machines, sharing the load. 

Reviewer 3: ```suggestion
In high volume scenarios, where there are more messages being produced than a single physical endpoint can handle, the logical endpoint can be [scaled out across multiple physical instances](/transports/scale-out.md) running on different machines, sharing the load. 
```

Each endpoint tracks [key performance metrics](/monitoring/metrics/definitions.md) that can be exposed as [Windows Performance Counters](/monitoring/metrics/performance-counters.md) or [collected into a central dashboard](/monitoring/metrics/in-servicepulse.md). [The monitoring demo](/tutorials/monitoring-demo/) demonstrates how to find performance bottlenecks and identify endpoints that are ready to scale out.

## Simple and testable

NServiceBus is designed with simplicity in mind. Message handlers don't need additional code to manage logging, exception handling, serialization, transactions, or the specifics of a queueing technology. This allows message handler code to focus on business logic.

> Long-running business workflows can be modeled in NServiceBus using a [saga](/nservicebus/sagas/). A saga is a C# class that handles a number of different messages over time, persisting its state between each step in the workflow. NServiceBus makes sure that each saga instance only processes a single message at a time, keeping its internal state consistent. The [NServiceBus sagas: Getting started](/nservicebus/sagas/) tutorial provides more details.

Reviewer 3: ```suggestion
Long-running business workflows can be modeled in NServiceBus using [sagas](/nservicebus/sagas/). A saga is a C# class that handles a number of different messages over time, persisting its state between each step in the workflow. NServiceBus makes sure that each saga instance only processes a single message at a time, keeping its internal state consistent. The [NServiceBus sagas: Getting started](/nservicebus/sagas/) tutorial provides more details.
```

Message handlers and sagas can be [tested in isolation](/nservicebus/testing/). Simulating an incoming message is as simple as creating a new C# message object and passing it to the appropriate handler or saga. The framework includes a suite of testing tools that capture the behavior of message handlers and sagas under test, allowing assertions to be made about them.

## Flexible

NServiceBus endpoints can be hosted anywhere code can be executed, such as in a Windows Service, a Docker container, or in the cloud with Azure or AWS. NServiceBus is compiled against .NET Standard 2.0 and endpoints can be run on [a variety of platforms](/nservicebus/upgrades/supported-platforms.md).  

> NServiceBus works with many different technology stacks, offering choices for [transport](/transports/) and [persistence](/persistence/). Out of the box, defaults are provided for [serialization](/nservicebus/serialization/), [dependency injection](/nservicebus/dependency-injection/), and [logging](/nservicebus/logging/). These defaults can be overridden if a sepcific technology is required. 

Reviewer 3: ```suggestion
NServiceBus works with many different technology stacks, offering choices for [transport](/transports/) and [persistence](/persistence/). Out of the box, defaults are provided for [serialization](/nservicebus/serialization/), [dependency injection](/nservicebus/dependency-injection/), and [logging](/nservicebus/logging/). These defaults can be overridden if a specific technology is required. 
```

## Particular Service Platform

NServiceBus is designed to work with the rest of the Particular Service Platform. All messages are instrumented with additional [headers](/nservicebus/messaging/headers.md) detailing key information about the message and how it was processed. As each message is processed it is forwarded to an [audit queue](/nservicebus/operations/auditing.md), where it is picked up by [ServiceControl](/servicecontrol/). [ServiceInsight](/serviceinsight/) connects to a ServiceControl instance to provide powerful visualizations of a running NServiceBus system, making it easy to understand message flows and timing.

> When a message fails to be processed, even after a number of retry strategies have been attempted, NServiceBus will forward the message to an [error queue](/nservicebus/recoverability/configure-error-handling.md) for manual investigation. Messages sent to the error queue are instrumented with headers containing details about the failure including a full exception stack trace. ServiceControl picks up messages from the error queue and makes them availible in [ServicePulse](/servicepulse/). Once the root cause of the failure has been found and corrected, all messages caused by the same problem can be retried at once. 

Reviewer 3: ```suggestion
When a message fails to be processed, even after a number of retry strategies have been attempted, NServiceBus will forward the message to an [error queue](/nservicebus/recoverability/configure-error-handling.md) for manual investigation. Messages sent to the error queue are instrumented with headers containing details about the failure including a full exception stack trace. ServiceControl picks up messages from the error queue and makes them available in [ServicePulse](/servicepulse/). Once the root cause of the failure has been found and corrected, all messages caused by the same problem can be retried at once. 
```
